### ISSUE TITLE
Figure out how to support "usable auth" for gcloud-node

### ISSUE THREAD

**Issue by jgeewax** [GoogleCloudPlatform/gcloud-node#376]
See https://github.com/google/google-api-nodejs-client/pull/354

Ideally we would somehow integrate this code. Maybe it means a separate google-authkit-node repository that we depend on ?

/cc @stephenplusplus @ryanseys

**Comment by ryanseys** [comment]
Ideally there would be 1 module that all of auth gets incorporated into. Right now, it's all mixed across `gtoken`, `google-service-account`, among other libraries and implementations. I like the way that `googleapis` currently accomplishes at least the accepting of generic auth objects:

``` js
var google  = require('googleapis');
var OAuth2 = google.auth.OAuth2Client;
var JWT = google.auth.JWT;
var myAuthClient = new OAuth2(/** ...  */);
myAuthClient = new JWT(/* ... */); // or use a JWT client
myAuthClient = 'SOME API KEY'; // or even a plain API key.

var drive = google.drive({ version: 'v2', auth: myAuthClient });
drive.files.insert(options, callback);
```

So I think it's a matter of removing that auth logic and objects (plus the changes in https://github.com/google/google-api-nodejs-client/pull/354) into a giant epic auth module.

**Comment by ryanseys** [comment]
There's a repo in the works here for nodejs auth: https://github.com/google/google-auth-library-nodejs

@jgeewax can you get @stephenplusplus access to this?

**Comment by stephenplusplus** [comment]
I'm in. Thanks!

**Comment by jgeewax** [comment]
Just to be clear, what I mean by usable auth here is "if I don't want to think about it in code, I don't have to think about it".

Right now, we have to explicitly define the credentials and whatnot somewhere. If that ever happens, we should use those settings. End of story.

If those aren't set anywhere, we should evaluate things in the following order:

1. If I have environment variables for anything credential related, use that. `break;`
1. If I'm in App Engine, use that set of credentials and defaults. `break;`
1. If I'm in Compute Engine, use that set of credentials and defaults. `break;`

This means that the "default case" (`var gcloud = require('gcloud');`) which previously meant "this still needs to be configured" now means "we'll try the list of stuff, and it might be configured depending on the the environment".

In code this means the following should work in GAE, GCE, or if I have the right environment variables set:

```javascript
var gcloud = require('gcloud');
dataset = gcloud.datastore.dataset();
query = dataset.createQuery('Person');
dataset.runQuery(query, function(...) { ... });
```

You can still *always* explicitly override those things later:

```javascript
// Continued from above, with the same gcloud that *was* configured properly via env
var dataset = gcloud.datastore.dataset({
  projectId: 'myProject',
  keyFilename: '/path/to/keyfile.json'
});
// Dataset uses another set of explicit credentials.
```

This may or may not merit a helper method `gcloud.isConfigured()` which returns true if and only if the instance of `gcloud` is ready to send requests.

**Comment by stephenplusplus** [comment]
This is a good plan, it's just a matter of nailing down the environment variables to look for in all of those environments and any API endpoints/authentication procedures that may changed based on the environment.

**Comment by ryanseys** [comment]
Here's how `googleapis` may be changed to support this: https://github.com/google/google-api-nodejs-client/pull/374

Adding here for reference when we want to consider using the same library.

**Comment by jgeewax** [comment]
Just FYI, hopefully we can rely on https://github.com/google/google-auth-library-nodejs (`npm install google-auth-library`).

Do the methods in that library do the thing we want ?

**Comment by tbetbetbe** [comment]
If https://github.com/google/google-auth-library-nodejs has gaps, please add an issue or send a PR.  We will respond promptly.

**Comment by anthmgoogle** [comment]
The new library https://github.com/google/google-auth-library-nodejs is defintely the home for this. The default credentials should by fully functional now. Please have Veneer take a dependency on this library. It should allow you to have a default behavior that uses the ADC under the covers and no credentials as input, and also have way of initializing the service that takes the type returned by this method as an optional input.

**Comment by ryanseys** [comment]
https://github.com/google/google-api-nodejs-client just shipped 2.0.0 with these changes!
